File: Old_xml_and_scripts/Jim/high_skills/new-18.5.2013/goToBall.rb
require 'high_skills/ruby_high_skill.rb'
include_class Java::sk.fiit.robocup.library.geometry.Angles
include_class Java::java.lang.Math

class GotoBall < RubyHighSkill

	def GotoBall.angle_range from, to
    	(from/180.0*Math::PI)..(to/180.0*Math::PI)
  end
  
  def distance(x,y)
    return Math.sqrt(x*x +  y*y)
  end

  @@rovno1 = angle_range(0.0, 25.0)
  @@rovno2 = angle_range(335.0, 360.0)
  @@vpravo = angle_range(270.0, 335.0)
  @@vlavo = angle_range(25.0, 90.0)
  @@vpravoZad = angle_range(180.0, 270.0)
  @@vlavoZad = angle_range(90.0, 180.0)
  
  @@daleko = 4
  @@blizko = 1.0
  @@tesne = 0.2
  @@odLopty = 0.7
  @@kLopte = 0.1

  
  def initialize(targetPosition)
    super()
    @tacticalInfo = Java::sk.fiit.jim.agent.models.TacticalInfo.getInstance
    @agentInfo = Java::sk.fiit.jim.agent.AgentInfo.getInstance
    @agentModel = Java::sk.fiit.jim.agent.models.AgentModel.getInstance
    
    @targetPosition = targetPosition
    @closeMove = false
    @indirectMove = false
    @finished = false
  end
  
  def pickLowSkill
    if @finished == true
      return get_skill("rollback")
    end
    
    @ballPosition = ball.position
    @agentPosition = @agentModel.getPosition()
    
    if @closeMove == false
      @movePosition = computeMidMovePoint()
    else
      @movePosition = computeFinalMovePoint()
    end
    @movePosition_phi = @movePosition.getPhi()
    
    
 #   puts "RM: " + @movePosition.getX().to_s+", "+@movePosition.getY().to_s
 #   puts "GM: " + @agentModel.globalize(@movePosition).getX().to_s+", "+@agentModel.globalize(@movePosition).getY().to_s
 #   puts "close: "+@closeMove.to_s+", indirect: "+@indirectMove.to_s+", finished: "+@finished.to_s
 #   puts "MOVE DISTANCE: "+Math.sqrt(@movePosition.getX()*@movePosition.getX() + @movePosition.getY()*@movePosition.getY()).to_s
 #   phi_degrees = @movePosition_phi * 180 / Math::PI
 #   puts "degrees: "+phi_degrees.to_s
 #   puts ""

    case
      when me.on_ground?
        puts "PADNUTY"
        @closeMove = false
        @indirectMove = false
        return
#       when @agentInfo.ballUnseen() > 10
#        puts "NEVIDIM"
#        return
      when distance(@movePosition.getX(), @movePosition.getY()) > @@daleko
        puts "DALEKO"
        @closeMove == false
        case
          when rovno?
            puts "ROVNO"
            return get_skill("walk_forward")
          when vlavo?
            puts "VLAVO"
            return get_skill("turn_left_36")
          when vpravo?
            puts "VPRAVO"
            return get_skill("turn_right_36")
          when vlavoZad?
            puts "VLAVOZAD"
          return get_skill("turn_left_36")
          when vpravoZad?
            puts "VPRAVOZAD"
            return get_skill("turn_right_36")
          else return get_skill("rollback")
        end
      when distance(@movePosition.getX(), @movePosition.getY()) > @@blizko
        puts "BLIZKO"
        case
          when rovno?
            puts "ROVNO"
            return get_skill("walk_forward")
          when vlavo?
            puts "VLAVO"
            return get_skill("turn_left_cont_20")
          when vpravo?
            puts "VPRAVO"
            return get_skill("turn_right_cont_20")
          when vlavoZad?
            puts "VLAVOZAD"
            return get_skill("turn_left_90")
          when vpravoZad?
            puts "VPRAVOZAD"
            return get_skill("turn_right_90")
          else return get_skill("rollback")
        end
      when (distance(@movePosition.getX(), @movePosition.getY()) > (@@tesne) && @indirectMove == true)
        puts "TESNE1"
        case
          when rovno?
            puts "ROVNO"
            return get_skill("walk_slow")
          when vlavo?
            puts "VLAVO"
            return get_skill("step_left")
          when vpravo?
            puts "VPRAVO"
            return get_skill("step_right")
          when vlavoZad?
            puts "VLAVOZAD"
            return get_skill("turn_left_90")
          when vpravoZad?
            puts "VPRAVOZAD"
            return get_skill("turn_right_90")
          else return get_skill("stojRovno")
        end
      when (distance(@movePosition.getX(), @movePosition.getY()) > (@@tesne) && @closeMove == false)
        puts "TESNE2"
        case
          when rovno?
            puts "ROVNO"
            return get_skill("walk_slow")
          when vlavo?
            puts "VLAVO"
            return get_skill("turn_left_cont_20")
          when vpravo?
            puts "VPRAVO"
            return get_skill("turn_right_cont_20")
          when vlavoZad?
            puts "VLAVOZAD"
            return get_skill("turn_left_90")
          when vpravoZad?
            puts "VPRAVOZAD"
            return get_skill("turn_right_90")
          else return get_skill("stojRovno")
        end
      when (distance(@movePosition.getX(), @movePosition.getY()) > @@tesne && @closeMove == true)
        puts "TESNE3"
        case
          when rovno?
            puts "ROVNO"
            return get_skill("walk_slow")
          when vlavo?
            puts "VLAVO"
            return get_skill("turn_left_cont_10")
          when vpravo?
            puts "VPRAVO"
            return get_skill("turn_right_cont_10")
          when vlavoZad?
            puts "VLAVOZAD"
            return get_skill("turn_left_90")
          when vpravoZad?
            puts "VPRAVOZAD"
            return get_skill("turn_right_90")
          else return get_skill("rollback")
        end
      when distance(@movePosition.getX(), @movePosition.getY()) <= @@tesne
      if @closeMove == true
 #       puts "NA MIESTE CLOSE"
 #       puts "A1: "+Angles.normalize(@agentPosition.getPhi()).to_s
 #       puts "A2: "+Angles.normalize(@targetPosition.getPhi()).to_s
 #       if (Angles.normalize(@agentPosition.getPhi()) - Angles.normalize(@targetPosition.getPhi())) > 0.3
 #         puts "ZLEEEE"
 #       end
        @finished = true
        return get_skill("step_left")
      elsif @closeMove == false
        puts "NA MIESTE"
        @closeMove = true;
        return get_skill("rollback")
 #     elsif @indirectMove == true
 #       puts "NA MIESTE INDIRECT"
 #       @indirectMove = false
 #       return get_skill("rollback")
      end
    else
      puts "FINAL ELSE"
      return get_skill("rollback")
    end
  end
  

  
   
  def rovno?
    @@rovno1.include? Angles.normalize(@movePosition_phi) or @@rovno2.include? Angles.normalize(@movePosition_phi)
  end
  
  def vlavo?
    @@vlavo.include? Angles.normalize(@movePosition_phi)
  end
 
  def vpravo?
    @@vpravo.include? Angles.normalize(@movePosition_phi)
  end
  
  def vlavoZad?
    @@vlavoZad.include? Angles.normalize(@movePosition_phi)
  end
  
  def vpravoZad?
    @@vpravoZad.include? Angles.normalize(@movePosition_phi)
  end
  

  
  
  
  
#  def back? target_position
#    Angles.angle_diff(target_phi, Math::PI) < Math::PI / 6.0
#  end
    
  
  
  
  

  def computeMidMovePoint()
#    dist = computeDistance(@targetPosition, @ballPosition)
    dist = (@targetPosition.getX() - @ballPosition.getX()).abs() + (@targetPosition.getY() - @ballPosition.getY()).abs()
    
    dX = (@targetPosition.getX() - @ballPosition.getX()) / dist
    dY = (@targetPosition.getY() - @ballPosition.getY()) / dist
    moveX = @ballPosition.getX() - dX * @@odLopty
    moveY = @ballPosition.getY() - dY * @@odLopty
    
    movePositionTmp = Java::sk.fiit.robocup.library.geometry.Vector3D.cartesian(moveX, moveY, 0)

    agentModel = Java::sk.fiit.jim.agent.models.AgentModel.getInstance()
    movePosition = agentModel.relativize(movePositionTmp)
    
    moveDist = movePositionTmp.getXYDistanceFrom(@agentPosition)
    ballDist = @ballPosition.getXYDistanceFrom(@agentPosition)
    
#    if ((@ballPosition.getPhi() - movePositionTmp.getPhi()).abs() < 0.1 && moveDist > ballDist)
#      puts "BP: "+@ballPosition.getX().to_s+", "+@ballPosition.getY().to_s
#      puts "MID: "+movePositionTmp.getX().to_s+", "+movePositionTmp.getY().to_s
#      if dX > dY
#        newMP = Java::sk.fiit.robocup.library.geometry.Vector3D.cartesian(
#                movePositionTmp.getX() + 1.5 * dY,
#                movePositionTmp.getY() - 1.5 * dX,
#                0)
#      else
#        newMP = Java::sk.fiit.robocup.library.geometry.Vector3D.cartesian(
#                movePositionTmp.getX() - 1.5 * dY,
#                movePositionTmp.getY() + 1.5 * dX,
#                0)
#      end
#      
#      movePosition = agentModel.relativize(newMP)
#      
#      @indirectMove = true
#    end
    
    return movePosition
  end
  
  def computeFinalMovePoint()
    dist = computeDistance(@targetPosition, @ballPosition)
        
    dX = (@targetPosition.getX() - @ballPosition.getX()) / dist
    dY = (@targetPosition.getY() - @ballPosition.getY()) / dist
    moveX = @ballPosition.getX() - dX * @@kLopte
    moveY = @ballPosition.getY() - dY * @@kLopte
    
    movePositionTmp = Java::sk.fiit.robocup.library.geometry.Vector3D.cartesian(moveX, moveY, 0)
    
    agentModel = Java::sk.fiit.jim.agent.models.AgentModel.getInstance()
    movePosition = agentModel.relativize(movePositionTmp)

    return movePosition
  end

  def computeDistance(pA, pB)
    distX = pA.getX() - pB.getX()
    distY = pA.getY() - pB.getY()
    return Math.sqrt(distX * distX + distY * distY)
  end

  def checkProgress
#    raise "Fallen!" if me.on_ground?
  end    
end

#in order not to defer the startup cost
GotoBall.new(@targetPosition)

File: Jim/src/sk/fiit/jim/agent/highskill/GotoBall.java
package sk.fiit.jim.agent.highskill;

import sk.fiit.jim.agent.AgentInfo;
import sk.fiit.jim.agent.models.AgentModel;
import sk.fiit.jim.agent.models.TacticalInfo;
import sk.fiit.jim.agent.models.WorldModel;
import sk.fiit.jim.agent.moves.LowSkill;
import sk.fiit.jim.agent.moves.LowSkills;
import sk.fiit.jim.agent.skills.HighSkill;
import sk.fiit.robocup.library.geometry.Angles;
import sk.fiit.robocup.library.geometry.Vector3D;

/**
 * 
 * Reimplemented from Ruby to Java by Stefan Horvath
 *
 */
public class GotoBall extends HighSkill {

	TacticalInfo tacticalInfo = TacticalInfo.getInstance();
	AgentInfo agentInfo = AgentInfo.getInstance();
	AgentModel agentModel = AgentModel.getInstance();
	
	Angles rovno1 = new Angles(0.0, 25.0);
	Angles rovno2 = new Angles(335.0, 360.0);
	Angles vpravo = new Angles(270.0, 335.0);
	Angles vlavo = new Angles(25.0, 90.0);
	Angles vpravoZad = new Angles(180.0, 270.0);
	Angles vlavoZad = new Angles(90.0, 180.0);
	
	double daleko = 4;
	double blizko = 1.0;
	double tesne = 0.2;
	double odLopty = 0.7;
	double kLopte = 0.1;
	
	Vector3D targetPosition = null;
	boolean closeMove = false;
	boolean indirectMove = false;
	boolean finished = false;
	Vector3D ballPosition = null;
	Vector3D agentPosition = null;
	double movePosition_phi;
		
	public GotoBall(Vector3D targetPosition) {
		super();
		this.targetPosition = targetPosition;
	}
	
	@Override
	public LowSkill pickLowSkill() {
		if (finished)
			return LowSkills.get("rollback");
		
		ballPosition =  WorldModel.getInstance().getBall().getPosition();
		agentPosition = agentModel.getPosition();
		
		Vector3D movePosition = null;
		if (!closeMove)
			movePosition = computeMidMovePoint();
		else
			movePosition = computeFinalMovePoint();
		movePosition_phi = movePosition.getPhi();
		
		if (agentModel.isOnGround()) {
			agentInfo.loguj("PADNUTY");
			closeMove = false;
			indirectMove = false;
			return null; // ????????
		}
		else if (distance(movePosition.getX(), movePosition.getY()) > daleko) {
			agentInfo.loguj("DALEKO");
			closeMove = false;
			if (rovno()) {
				agentInfo.loguj("ROVNO");
				return LowSkills.get("walk_forward");
			}
			else if (vlavo()) {
				agentInfo.loguj("VLAVO");
				return LowSkills.get("turn_left_36");
			}
			else if (vpravo()) {
				agentInfo.loguj("VPRAVO");
				return LowSkills.get("turn_right_36");
			}
			else if (vlavoZad()) {
				agentInfo.loguj("VLAVOZAD");
				return LowSkills.get("turn_left_36");
			}
			else if (vpravoZad()) {
				agentInfo.loguj("VPRAVOZAD");
				return LowSkills.get("turn_right_36");
			}
			else {
				return LowSkills.get("rollback");
			}
		}
		else if (distance(movePosition.getX(), movePosition.getY()) > blizko) {
			agentInfo.loguj("BLIZKO");
			if (rovno()) {
				agentInfo.loguj("ROVNO");
				return LowSkills.get("walk_forward");
			}
			else if (vlavo()) {
				agentInfo.loguj("VLAVO");
				return LowSkills.get("turn_left_cont_20");
			}
			else if (vpravo()) {
				agentInfo.loguj("VPRAVO");
				return LowSkills.get("turn_right_cont_20");
			}
			else if (vlavoZad()) {
				agentInfo.loguj("VLAVOZAD");
				return LowSkills.get("turn_left_90");
			}
			else if (vpravoZad()) {
				agentInfo.loguj("VPRAVOZAD");
				return LowSkills.get("turn_right_90");
			}
			else {
				return LowSkills.get("rollback");
			}
		}
		else if ((distance(movePosition.getX(), movePosition.getY()) > tesne) && indirectMove) {
			agentInfo.loguj("TESNE1");
			if (rovno()) {
				agentInfo.loguj("ROVNO");
				return LowSkills.get("walk_slow");
			}
			else if (vlavo()) {
				agentInfo.loguj("VLAVO");
				return LowSkills.get("step_left");
			}
			else if (vpravo()) {
				agentInfo.loguj("VPRAVO");
				return LowSkills.get("step_right");
			}
			else if (vlavoZad()) {
				agentInfo.loguj("VLAVOZAD");
				return LowSkills.get("turn_left_90");
			}
			else if (vpravoZad()) {
				agentInfo.loguj("VPRAVOZAD");
				return LowSkills.get("turn_right_90");
			}
			else {
				return LowSkills.get("stojRovno");
			}
		}
		else if ((distance(movePosition.getX(), movePosition.getY()) > tesne) && !closeMove) {
			agentInfo.loguj("TESNE2");
			if (rovno()) {
				agentInfo.loguj("ROVNO");
				return LowSkills.get("walk_slow");
			}
			else if (vlavo()) {
				agentInfo.loguj("VLAVO");
				return LowSkills.get("turn_left_cont_20");
			}
			else if (vpravo()) {
				agentInfo.loguj("VPRAVO");
				return LowSkills.get("turn_right_cont_20");
			}
			else if (vlavoZad()) {
				agentInfo.loguj("VLAVOZAD");
				return LowSkills.get("turn_left_90");
			}
			else if (vpravoZad()) {
				agentInfo.loguj("VPRAVOZAD");
				return LowSkills.get("turn_right_90");
			}
			else {
				return LowSkills.get("stojRovno");
			}
		}
		else if ((distance(movePosition.getX(), movePosition.getY()) > tesne) && closeMove) {
			agentInfo.loguj("TESNE3");
			if (rovno()) {
				agentInfo.loguj("ROVNO");
				return LowSkills.get("walk_slow");
			}
			else if (vlavo()) {
				agentInfo.loguj("VLAVO");
				return LowSkills.get("turn_left_cont_10");
			}
			else if (vpravo()) {
				agentInfo.loguj("VPRAVO");
				return LowSkills.get("turn_right_cont_10");
			}
			else if (vlavoZad()) {
				agentInfo.loguj("VLAVOZAD");
				return LowSkills.get("turn_left_90");
			}
			else if (vpravoZad()) {
				agentInfo.loguj("VPRAVOZAD");
				return LowSkills.get("turn_right_90");
			}
			else {
				return LowSkills.get("rollback");
			}
		}
		else if (distance(movePosition.getX(), movePosition.getY()) <= tesne) {
			if (closeMove) {
				finished = true;
				return LowSkills.get("step_left");
			}
			else {
				agentInfo.loguj("NA MIESTE");
				closeMove = true;
				return LowSkills.get("rollback");
			}
		}
		else {
			agentInfo.loguj("FINAL ELSE");
			return LowSkills.get("rollback");
		}
	}
	
	public boolean rovno() {
		return (rovno1.include(Angles.normalize(movePosition_phi)) || rovno2.include(Angles.normalize(movePosition_phi)));
	}
	
	public boolean vlavo() {
		return vlavo.include(Angles.normalize(movePosition_phi));
	}
	
	public boolean vpravo() {
		return vpravo.include(Angles.normalize(movePosition_phi));
	}
	
	public boolean vlavoZad() {
		return vlavoZad.include(Angles.normalize(movePosition_phi));
	}
	
	public boolean vpravoZad() {
		return vpravoZad.include(Angles.normalize(movePosition_phi));
	}


	public Vector3D computeMidMovePoint() {
		double dist = Math.abs(targetPosition.getX() - ballPosition.getX()) + Math.abs(targetPosition.getY() - ballPosition.getY());
		double dX = (targetPosition.getX() - ballPosition.getX()) / dist;
		double moveX = ballPosition.getX() - dX * odLopty;
		double moveY = ballPosition.getY() - dX * odLopty;
		Vector3D movePositionTmp = Vector3D.cartesian(moveX, moveY, 0);
		Vector3D movePosition = agentModel.relativize(movePositionTmp);
		
		return movePosition;
	}
	
	public Vector3D computeFinalMovePoint() {
		double dist = computeDistance(targetPosition, ballPosition);
		double dX = (targetPosition.getX() - ballPosition.getX()) / dist;
		double dY = (targetPosition.getY() - ballPosition.getY()) / dist;
		double moveX = ballPosition.getX() - dX * kLopte;
		double moveY = ballPosition.getY() - dY * kLopte;
		
		Vector3D movePositionTmp = Vector3D.cartesian(moveX, moveY, 0);
		Vector3D movePosition = agentModel.relativize(movePositionTmp);
		return movePosition;
	}
	
	public double computeDistance(Vector3D pA, Vector3D pB) {
		double distX = pA.getX() - pB.getX();
		double distY = pA.getY() - pB.getY();
		return Math.sqrt(distX * distX + distY * distY);
	}
	
	public double distance(double x, double y) {
		return Math.sqrt(x*x + y*y);
	}
	
	@Override
	public void checkProgress() throws Exception {
	}

}
